worker/controlsocket/handlers.go

```diff
@@ -0,0 +1,237 @@
+	"github.com/juju/juju/state"
+)
+
+const (
+	// jujuMetricsUserPrefix defines the "namespace" in which this worker is
+	// allowed to create/remove users.
+	jujuMetricsUserPrefix = "juju-metrics-"
+
+	// userCreator is the listed "creator" of metrics users in state.
+	userCreator = "controlsocket"
+)
+
+func (w *Worker) registerHandlers(handle func(pattern string, handler http.Handler)) {
+	// We're using a fully REST-ful API structure here.
+	handle("/metrics-users", http.HandlerFunc(w.handleAddMetricsUser))
+	handle("/metrics-users/", http.HandlerFunc(w.handleRemoveMetricsUser))
+}
+
+func (w *Worker) handleAddMetricsUser(resp http.ResponseWriter, req *http.Request) {
+	if req.Method != http.MethodPost {
+		resp.Header().Add("Allow", http.MethodPost)
+		w.writeResponse(resp, http.StatusMethodNotAllowed, errorResp("method must be %q, got %q", http.MethodPost, req.Method))
+		return
+	}
+
+	body := &struct {
+		Username string `json:"username"`
+		Password string `json:"password"`
+	}{}
+
+	defer req.Body.Close()
+	err := json.NewDecoder(req.Body).Decode(body)
+	if errors.Is(err, io.EOF) {
+		w.writeResponse(resp, http.StatusBadRequest, errorResp("missing request body"))
+		return
+	} else if err != nil {
+		w.writeResponse(resp, http.StatusBadRequest, errorResp("request body is not valid JSON: %v", err.Error()))
+		return
+	}
+
+	err = w.addMetricsUser(body.Username, body.Password)
+	if err != nil {
+		w.writeResponse(resp, asHTTPCode(err), errorResp(err.Error()))
+		return
+	}
+	w.writeResponse(resp, http.StatusOK, struct{}{})
+}
+
+func (w *Worker) addMetricsUser(username, password string) error {
+	err := validateMetricsUsername(username)
+	if err != nil {
+		return err
+	}
+
+	if password == "" {
+		return errors.NotValidf("empty password")
+	}
+
+	user, err := w.State.AddUser(username, username, password, userCreator)
+	if errors.Is(err, errors.AlreadyExists) {
+		if user.CreatedBy() != userCreator {
+			// We shouldn't mess with users that have not been created by us.
+			return errors.AlreadyExistsf("user %q created by %q", user.Name(), user.CreatedBy())
+		}
+	} else if err != nil {
+		return errors.Annotatef(err, "failed to create user %q: %v", username, err)
+	}
+	userTag := user.UserTag()
+
+	// At this point, the operation is in a partially completed state - we've
+	// added the user, but haven't granted them the correct model permissions.
+	// If there is an error granting permissions, we should attempt to "rollback"
+	// and remove the user again.
+	defer func() {
+		if err != nil {
+			err := w.State.RemoveUser(userTag)
+			if err != nil {
+				// Best we can do here is log an error.
+				w.Logger.Warningf("add metrics user failed, but could not clean up user %q: %v",
+					username, err)
+			}
+		}
+	}()
+
+	var model model
+	model, err = w.State.Model()
+	if err != nil {
+		return errors.Annotatef(err, "retrieving current model: %v", err)
+	}
+
+	_, err = model.AddUser(state.UserAccessSpec{
```

Comment:
What happens if the model user already exists?

Reply:
An example of a failure mode we could already have with the current code:
- client requests add user foo
- server adds user foo and adds model user foo
- server crashes/client doesn't receive response etc
- client tries again. add user foo. 
- server sees state user already exists ok (maybe we partially failed?), try add model user, oh it already exists.
- server deletes user, even though the previous request succeeded
- server says error already exists.. client says 🤷‍♂️